src/linux/py_proc.h

```diff
@@ -81,9 +81,7 @@ _py_proc__analyze_elf64(py_proc_t * self) {
 
   if (ehdr.e_phnum) {
     Elf64_Phdr * phdr = (Elf64_Phdr *) (elf_map + ehdr.e_phoff);
-    base = phdr->p_vaddr - phdr->p_offset;
-    if (!base)
-      base = (Elf64_Addr) self->map.elf.base;
+    base = ((Elf64_Addr) phdr->p_vaddr) / phdr->p_align * phdr->p_align;
```

Comment:
Actually to be corerct, we should take the min over all the `PT_LOAD` entries.

Reply:
I am now tempted to say that this should just be
`base = (Elf64_Addr) self->map.elf.base;`
in every case.

Reply:
This is trying to implement:

> truncating the memory address to the nearest multiple of the maximum page size.

but I have never seen an ELF file where the virtual address is not already aligned to the page size.

Maybe something like this would be clearer:

```
 base = ((Elf64_Addr) phdr->p_vaddr) - (Elf64_Addr) phdr->p_vaddr % phdr->p_align;
```

Reply:
In this code, `base` is now the base address **in the ELF file**. This it typically `0x0` for `ET_DYN` files.

However, `self->map.elf.base` is the **runtime** base address. Typically something like `0x55e5f5b67000`.

Reply:
In fact, you might be right that one actually needs the value of the base from the maps file minus the computed base. One then adds this value to `st_value` to get the symbol address. I also think that the address is probably aligned to the page boundary and that one can take the virtual address of the first PT_LOAD segment as base address.

Reply:
> In this code, `base` is now the base address **in the ELF file**. This it typically `0x0` for `ET_DYN` files.
> 
> However, `self->map.elf.base` is the **runtime** base address. Typically something like `0x55e5f5b67000`.

I agree. I then think that the ternary operator is not required, as `self->map.elf.base + sym->st_value - base`, with `base` being the ELF base is good in all cases. However, I have the feeling that the right formula is
`self->map.elf.base + offset + sym->st_value - base`,
where `offset` is the `p_offset` of the segment with the lowest `p_vaddr` (I believe it is the first `PT_LOAD` segment in the ELF).

Reply:
> (I believe it is the first PT_LOAD segment in the ELF).

The ELF spec says :

> PT_LOAD: Loadable segment entries in the programheader table appear in ascending order, sorted on thep_vaddrmember.

So yes, the first `PT_LOAD` should always be the lowest.